File: packages/python/src/refract/otel.py
"""OpenTelemetry OTLP/HTTP JSON bridge; no OTel runtime dependency is required."""

from __future__ import annotations

import hashlib
import json
import re
import urllib.request
from datetime import UTC, datetime
from typing import Any

from . import SPEC_VERSION, _snapshot


def _id(value: str, length: int) -> str:
    return hashlib.sha256(value.encode()).hexdigest()[:length]


def _ns(value: str) -> int:
    return int(datetime.fromisoformat(value.replace("Z", "+00:00")).timestamp() * 1_000_000_000)


def _time(value: int) -> str:
    return datetime.fromtimestamp(value / 1_000_000_000, UTC).isoformat()


def _value(value: Any) -> dict:
    if isinstance(value, bool):
        return {"boolValue": value}
    if isinstance(value, int):
        return {"intValue": str(value)}
    if isinstance(value, float):
        return {"doubleValue": value}
    if isinstance(value, str):
        return {"stringValue": value}
    if isinstance(value, (list, tuple)):
        return {"arrayValue": {"values": [_value(item) for item in value]}}
    if isinstance(value, dict):
        return {"kvlistValue": {"values": _attributes(value)}}
    return {}


def _attributes(values: dict) -> list[dict]:
    return [{"key": key, "value": _value(value)} for key, value in values.items()]


def _decode(value: dict) -> Any:
    for key, cast in (
        ("stringValue", str),
        ("boolValue", bool),
        ("intValue", int),
        ("doubleValue", float),
    ):
        if key in value:
            return cast(value[key])
    if "arrayValue" in value:
        return [_decode(item) for item in value["arrayValue"].get("values", [])]
    if "kvlistValue" in value:
        return _decode_attributes(value["kvlistValue"].get("values", []))
    return None


def _decode_attributes(values: list) -> dict:
    return {item["key"]: _decode(item["value"]) for item in values}


def to_otlp(recording: dict, *, service_name: str = "llm-refract") -> dict:
    """Encode one run as OTLP JSON traces, preserving event payloads in namespaced attributes."""
    recording = _snapshot(recording)
    trace_id = _id(recording["id"], 32)
    spans = []
    for event in recording["events"]:
        start = _ns(event["timestamp"])
        attributes = {
            "refract.event.id": event["id"],
            "refract.event.type": event["type"],
            "refract.event.input": json.dumps(event.get("input"), ensure_ascii=False),
            "refract.event.output": json.dumps(event.get("output"), ensure_ascii=False),
            "refract.event.attributes": json.dumps(event.get("attributes", {}), ensure_ascii=False),
            "refract.event.replay_policy": event["replay_policy"],
            "refract.event.status": event["status"],
        }
        for source, target in {
            "model": "gen_ai.request.model",
            "provider": "gen_ai.provider.name",
            "input_tokens": "gen_ai.usage.input_tokens",
            "output_tokens": "gen_ai.usage.output_tokens",
        }.items():
            if source in event["attributes"]:
                attributes[target] = event["attributes"][source]
        span = {
            "traceId": trace_id,
            "spanId": _id(event["id"], 16),
            "name": event["name"],
            "kind": 3 if event["type"] == "generation" else 1,
            "startTimeUnixNano": str(start),
            "endTimeUnixNano": str(start + int(event["duration_ms"] * 1_000_000)),
            "attributes": _attributes(attributes),
            "status": {"code": 2 if event["status"] == "failed" else 1},
        }
        if event.get("parent_id"):
            span["parentSpanId"] = _id(event["parent_id"], 16)
        spans.append(span)
    resource = {
        "service.name": service_name,
        "refract.run.id": recording["id"],
        "refract.run.name": recording["name"],
        "refract.run.metadata": json.dumps(recording.get("metadata", {})),
        "refract.run.started_at": recording["started_at"],
        "refract.run.ended_at": recording.get("ended_at"),
        "refract.run.status": recording["status"],
    }
    return {
        "resourceSpans": [
            {
                "resource": {"attributes": _attributes(resource)},
                "scopeSpans": [{"scope": {"name": "llm-refract"}, "spans": spans}],
            }
        ]
    }


def from_otlp(document: dict) -> list[dict]:
    """Import one canonical run per trace, sorting parents before children.

    External parents remain in otel.external_parent_id. Unsupported OTel span events
    and links are retained under attributes; no application code is executed.
    """
    groups: dict[str, list] = {}
    for resource_spans in document.get("resourceSpans", []):
        resource = _decode_attributes(resource_spans.get("resource", {}).get("attributes", []))
        for scope in resource_spans.get("scopeSpans", []):
            for span in scope.get("spans", []):
                trace_id, span_id = span.get("traceId", ""), span.get("spanId", "")
                if not re.fullmatch(r"[0-9a-fA-F]{32}", trace_id) or int(trace_id, 16) == 0:
                    raise ValueError("invalid OTLP trace ID")
                if not re.fullmatch(r"[0-9a-fA-F]{16}", span_id) or int(span_id, 16) == 0:
                    raise ValueError("invalid OTLP span ID")
                item = dict(span, spanId=span_id.lower())
                if item.get("parentSpanId"):
                    item["parentSpanId"] = item["parentSpanId"].lower()
                groups.setdefault(trace_id.lower(), []).append((item, resource))
    runs = []
    for trace_id, entries in groups.items():
        resource = entries[0][1]
        run_id = resource.get("refract.run.id", f"run_otel_{trace_id}")
        starts = [int(item[0]["startTimeUnixNano"]) for item in entries]
        ends = [int(item[0]["endTimeUnixNano"]) for item in entries]
        run = {
            "spec_version": SPEC_VERSION,
            "id": run_id,
            "name": resource.get("refract.run.name", resource.get("service.name", "OTel trace")),
            "started_at": resource.get("refract.run.started_at", _time(min(starts))),
            "ended_at": resource.get("refract.run.ended_at", _time(max(ends))),
            "status": resource.get("refract.run.status", "completed"),
            "metadata": json.loads(resource.get("refract.run.metadata", "{}")),
            "events": [],
        }
        mapped = {}
        for span, _ in entries:
            if span["spanId"] in mapped:
                raise ValueError("duplicate OTLP span ID within trace")
            attrs = _decode_attributes(span.get("attributes", []))
            mapped[span["spanId"]] = attrs.get("refract.event.id", f"evt_{span['spanId']}")
        if len(set(mapped.values())) != len(mapped):
            raise ValueError("duplicate mapped Refract event IDs")
        pending = list(entries)
        emitted: set[str] = set()
        while pending:
            previous_length = len(pending)
            for span, span_resource in list(pending):
                parent = span.get("parentSpanId")
                if parent in mapped and parent not in emitted:
                    continue
                attributes = _decode_attributes(span.get("attributes", []))
                values = json.loads(attributes.get("refract.event.attributes", "{}"))
                for source, target in {
                    "gen_ai.request.model": "model",
                    "gen_ai.provider.name": "provider",
                    "gen_ai.system": "provider",
                    "gen_ai.usage.input_tokens": "input_tokens",
                    "gen_ai.usage.output_tokens": "output_tokens",
                }.items():
                    if source in attributes:
                        values[target] = attributes[source]
                if "refract.event.id" not in attributes:
                    values["otel.attributes"] = attributes
                    values["otel.resource"] = span_resource
                if parent and parent not in mapped:
                    values["otel.external_parent_id"] = parent
                for field in ("events", "links"):
                    if field in span:
                        values[f"otel.{field}"] = span[field]
                start, end = int(span["startTimeUnixNano"]), int(span["endTimeUnixNano"])
                if end < start:
                    raise ValueError("OTLP span ends before it starts")
                failed = span.get("status", {}).get("code") in (2, "STATUS_CODE_ERROR")
                if failed:
                    run["status"] = "failed"
                run["events"].append(
                    {
                        "id": mapped[span["spanId"]],
                        "run_id": run_id,
                        "parent_id": mapped.get(parent),
                        "type": attributes.get(
                            "refract.event.type", "generation" if "model" in values else "decision"
                        ),
                        "name": span["name"],
                        "timestamp": _time(start),
                        "duration_ms": (end - start) / 1_000_000,
                        "status": attributes.get(
                            "refract.event.status", "failed" if failed else "completed"
                        ),
                        "input": json.loads(attributes.get("refract.event.input", "null")),
                        "output": json.loads(attributes.get("refract.event.output", "null")),
                        "attributes": values,
                        "replay_policy": attributes.get("refract.event.replay_policy", "RECORDED"),
                    }
                )
                emitted.add(span["spanId"])
                pending.remove((span, span_resource))
            if len(pending) == previous_length:
                raise ValueError("OTLP spans contain a parent cycle")
        runs.append(_snapshot(run))
    return runs


def from_spans(spans) -> list[dict]:
    """Convert completed OpenTelemetry SDK ReadableSpans without an OTLP dependency.

    Supply a complete trace where possible. This function only imports the supplied
    spans: it does not reconstruct missing parents or join different export batches.
    """
    resource_spans = []
    for source in spans:
        context = source.get_span_context()
        if context is None or source.start_time is None or source.end_time is None:
            raise ValueError("OTel SDK spans must have context and completed timestamps")
        item = {
            "traceId": format(context.trace_id, "032x"),
            "spanId": format(context.span_id, "016x"),
            "name": source.name,
            "startTimeUnixNano": str(source.start_time),
            "endTimeUnixNano": str(source.end_time),
            "attributes": _attributes(dict(source.attributes or {})),
            "status": {"code": source.status.status_code.value},
            "events": [
                {
                    "name": event.name,
                    "timeUnixNano": str(event.timestamp),
                    "attributes": _attributes(dict(event.attributes or {})),
                }
                for event in source.events
            ],
            "links": [
                {
                    "traceId": format(link.context.trace_id, "032x"),
                    "spanId": format(link.context.span_id, "016x"),
                    "attributes": _attributes(dict(link.attributes or {})),
                }
                for link in source.links
            ],
        }
        if source.parent:
            item["parentSpanId"] = format(source.parent.span_id, "016x")
        resource_spans.append(
            {
                "resource": {"attributes": _attributes(dict(source.resource.attributes))},
                "scopeSpans": [{"spans": [item]}],
            }
        )
    return from_otlp({"resourceSpans": resource_spans})


def export_otlp(recording: dict, endpoint: str, *, headers: dict | None = None, timeout=10) -> None:
    """Synchronously POST OTLP JSON to an OTel collector's /v1/traces endpoint."""
    _send_otlp(to_otlp(recording), endpoint, headers=headers, timeout=timeout)


def _send_otlp(document: dict, endpoint: str, *, headers: dict | None = None, timeout=10) -> None:
    address = endpoint.rstrip("/")
    if not address.endswith("/v1/traces"):
        address += "/v1/traces"
    request = urllib.request.Request(
        address,
        json.dumps(document, allow_nan=False).encode(),
        {"Content-Type": "application/json", **(headers or {})},
        method="POST",
    )
    with urllib.request.urlopen(request, timeout=timeout) as response:
        body = response.read()
    acknowledgement = json.loads(body) if body else {}
    partial = acknowledgement.get("partialSuccess") or {}
    if int(partial.get("rejectedSpans", 0)) or partial.get("errorMessage"):
        raise RuntimeError("OTLP collector reported partial rejection")

File: packages/typescript/src/otel.ts
import { createHash } from "node:crypto";
import {
  redact,
  type Execution,
  type ExecutionEvent,
  type Json,
} from "./index.js";

type Data = Record<string, unknown>;
function object(value: unknown): Data {
  if (!value || typeof value !== "object" || Array.isArray(value))
    throw new Error("Expected an OTLP object");
  return value as Data;
}
function array(value: unknown): unknown[] {
  if (value === undefined) return [];
  if (!Array.isArray(value)) throw new Error("Expected an OTLP array");
  return value;
}
function hash(value: string, length: number): string {
  return createHash("sha256").update(value).digest("hex").slice(0, length);
}
function value(input: Json): Data {
  if (input === null) return {};
  if (typeof input === "boolean") return { boolValue: input };
  if (typeof input === "number")
    return Number.isSafeInteger(input)
      ? { intValue: String(input) }
      : { doubleValue: input };
  if (typeof input === "string") return { stringValue: input };
  if (Array.isArray(input)) return { arrayValue: { values: input.map(value) } };
  return { kvlistValue: { values: attributes(input) } };
}
function attributes(input: Record<string, Json>): Data[] {
  return Object.entries(input).map(([key, item]) => ({
    key,
    value: value(item),
  }));
}
function decode(input: unknown): Json {
  const item = object(input);
  if ("stringValue" in item) return String(item.stringValue);
  if ("boolValue" in item) return Boolean(item.boolValue);
  if ("intValue" in item) {
    const result = Number(item.intValue);
    return Number.isSafeInteger(result) ? result : String(item.intValue);
  }
  if ("doubleValue" in item) return Number(item.doubleValue);
  if ("arrayValue" in item)
    return array(object(item.arrayValue).values).map(decode);
  if ("kvlistValue" in item)
    return decodeAttributes(object(item.kvlistValue).values);
  return null;
}
function decodeAttributes(input: unknown): Record<string, Json> {
  return Object.fromEntries(
    array(input).map((entry) => {
      const item = object(entry);
      return [String(item.key), decode(item.value)];
    }),
  );
}
function nanos(timestamp: string): bigint {
  const ms = Date.parse(timestamp);
  if (!Number.isFinite(ms)) throw new Error("Invalid execution timestamp");
  return BigInt(ms) * 1_000_000n;
}
function timestamp(nano: unknown): string {
  return new Date(Number(BigInt(String(nano)) / 1_000_000n)).toISOString();
}
const semanticKeys: Record<string, string> = {
  model: "gen_ai.request.model",
  provider: "gen_ai.provider.name",
  input_tokens: "gen_ai.usage.input_tokens",
  output_tokens: "gen_ai.usage.output_tokens",
};
/** Encode a completed execution as OTLP/HTTP JSON traces without an OTel dependency. */
export function toOtlp(
  recording: Execution,
  serviceName = "llm-refract",
): Data {
  const run = redact(
    structuredClone(recording) as unknown as Json,
  ) as unknown as Execution;
  const spans = run.events.map((event) => {
    const start = nanos(event.timestamp);
    const attrs: Record<string, Json> = {
      "refract.event.id": event.id,
      "refract.event.type": event.type,
      "refract.event.input": JSON.stringify(event.input ?? null),
      "refract.event.output": JSON.stringify(event.output ?? null),
      "refract.event.attributes": JSON.stringify(event.attributes ?? {}),
      "refract.event.replay_policy": event.replay_policy ?? "RECORDED",
    };
    for (const [source, target] of Object.entries(semanticKeys))
      if (event.attributes?.[source] !== undefined)
        attrs[target] = event.attributes[source];
    return {
      traceId: hash(run.id, 32),
      spanId: hash(event.id, 16),
      ...(event.parent_id ? { parentSpanId: hash(event.parent_id, 16) } : {}),
      name: event.name,
      kind: event.type === "generation" ? 3 : 1,
      startTimeUnixNano: String(start),
      endTimeUnixNano: String(
        start + BigInt(Math.round((event.duration_ms ?? 0) * 1_000_000)),
      ),
      attributes: attributes(attrs),
      status: { code: event.status === "failed" ? 2 : 1 },
    };
  });
  return {
    resourceSpans: [
      {
        resource: {
          attributes: attributes({
            "service.name": serviceName,
            "refract.run.id": run.id,
            "refract.run.name": run.name,
            "refract.run.metadata": JSON.stringify(run.metadata),
            "refract.run.started_at": run.started_at,
            "refract.run.ended_at": run.ended_at,
            "refract.run.status": run.status,
          }),
        },
        scopeSpans: [{ scope: { name: "llm-refract" }, spans }],
      },
    ],
  };
}
/** Import one execution per trace, keeping external parent IDs, links and span events. */
export function fromOtlp(document: unknown): Execution[] {
  const groups = new Map<
    string,
    { span: Data; resource: Record<string, Json> }[]
  >();
  for (const resource of array(object(document).resourceSpans)) {
    const item = object(resource);
    const attrs = decodeAttributes(object(item.resource ?? {}).attributes);
    for (const scope of array(item.scopeSpans))
      for (const rawSpan of array(object(scope).spans)) {
        const span = { ...object(rawSpan) };
        const traceId = String(span.traceId).toLowerCase();
        const spanId = String(span.spanId).toLowerCase();
        if (!/^[a-f0-9]{32}$/.test(traceId) || /^0+$/.test(traceId))
          throw new Error("Invalid OTLP trace ID");
        if (!/^[a-f0-9]{16}$/.test(spanId) || /^0+$/.test(spanId))
          throw new Error("Invalid OTLP span ID");
        span.spanId = spanId;
        if (span.parentSpanId)
          span.parentSpanId = String(span.parentSpanId).toLowerCase();
        const entries = groups.get(traceId) ?? [];
        entries.push({ span, resource: attrs });
        groups.set(traceId, entries);
      }
  }
  return [...groups].map(([traceId, entries]) => {
    const resource = entries[0].resource;
    const starts = entries.map(({ span }) =>
      BigInt(String(span.startTimeUnixNano)),
    );
    const ends = entries.map(({ span }) =>
      BigInt(String(span.endTimeUnixNano)),
    );
    const run: Execution = {
      spec_version: "refract.execution.v1",
      id: String(resource["refract.run.id"] ?? `run_otel_${traceId}`),
      name: String(
        resource["refract.run.name"] ??
          resource["service.name"] ??
          "OTel trace",
      ),
      status: (resource["refract.run.status"] ??
        "completed") as Execution["status"],
      started_at: String(
        resource["refract.run.started_at"] ??
          timestamp(starts.reduce((a, b) => (a < b ? a : b))),
      ),
      ended_at:
        resource["refract.run.ended_at"] === null
          ? null
          : String(
              resource["refract.run.ended_at"] ??
                timestamp(ends.reduce((a, b) => (a > b ? a : b))),
            ),
      metadata: object(
        JSON.parse(String(resource["refract.run.metadata"] ?? "{}")),
      ) as Record<string, Json>,
      events: [],
    };
    const mapped = new Map<string, string>();
    for (const { span } of entries) {
      const id = String(span.spanId);
      if (mapped.has(id)) throw new Error("Duplicate OTLP span ID");
      mapped.set(
        id,
        String(
          decodeAttributes(span.attributes)["refract.event.id"] ?? `evt_${id}`,
        ),
      );
    }
    if (new Set(mapped.values()).size !== mapped.size)
      throw new Error("Duplicate mapped event IDs");
    const pending = [...entries];
    const emitted = new Set<string>();
    while (pending.length) {
      let changed = false;
      for (const entry of [...pending]) {
        const { span, resource: spanResource } = entry;
        const parent = span.parentSpanId
          ? String(span.parentSpanId)
          : undefined;
        if (parent && mapped.has(parent) && !emitted.has(parent)) continue;
        const attrs = decodeAttributes(span.attributes);
        const values = object(
          JSON.parse(String(attrs["refract.event.attributes"] ?? "{}")),
        ) as Record<string, Json>;
        for (const [target, source] of Object.entries(semanticKeys))
          if (attrs[source] !== undefined) values[target] = attrs[source];
        if (
          values.provider === undefined &&
          attrs["gen_ai.system"] !== undefined
        )
          values.provider = attrs["gen_ai.system"];
        if (attrs["refract.event.id"] === undefined) {
          values["otel.attributes"] = attrs;
          values["otel.resource"] = spanResource;
        }
        if (parent && !mapped.has(parent))
          values["otel.external_parent_id"] = parent;
        for (const field of ["events", "links"])
          if (span[field] !== undefined)
            values[`otel.${field}`] = span[field] as Json;
        const start = BigInt(String(span.startTimeUnixNano));
        const end = BigInt(String(span.endTimeUnixNano));
        if (end < start) throw new Error("OTLP span ends before it starts");
        const status = object(span.status ?? {}).code;
        const failed = status === 2 || status === "STATUS_CODE_ERROR";
        if (failed) run.status = "failed";
        run.events.push({
          id: mapped.get(String(span.spanId))!,
          run_id: run.id,
          parent_id: parent ? (mapped.get(parent) ?? null) : null,
          type: (attrs["refract.event.type"] ??
            (values.model
              ? "generation"
              : "decision")) as ExecutionEvent["type"],
          name: String(span.name),
          timestamp: timestamp(start),
          duration_ms: Number(end - start) / 1_000_000,
          status: failed ? "failed" : "completed",
          input: JSON.parse(
            String(attrs["refract.event.input"] ?? "null"),
          ) as Json,
          output: JSON.parse(
            String(attrs["refract.event.output"] ?? "null"),
          ) as Json,
          attributes: values,
          replay_policy: (attrs["refract.event.replay_policy"] ??
            "RECORDED") as ExecutionEvent["replay_policy"],
        });
        emitted.add(String(span.spanId));
        pending.splice(pending.indexOf(entry), 1);
        changed = true;
      }
      if (!changed) throw new Error("OTLP parent cycle");
    }
    return redact(run as unknown as Json) as unknown as Execution;
  });
}
/** Explicit synchronous-in-flow export; pass the collector base URL or /v1/traces. */
export async function exportOtlp(
  run: Execution,
  endpoint: string,
  options: { headers?: Record<string, string>; timeoutMs?: number } = {},
): Promise<void> {
  const base = endpoint.replace(/\/$/, "");
  const response = await fetch(
    base.endsWith("/v1/traces") ? base : `${base}/v1/traces`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json", ...options.headers },
      body: JSON.stringify(toOtlp(run)),
      signal: AbortSignal.timeout(options.timeoutMs ?? 10_000),
    },
  );
  if (!response.ok) throw new Error(`OTLP export failed: ${response.status}`);
  const result = object(await response.json());
  const partial = object(result.partialSuccess ?? {});
  if (BigInt(String(partial.rejectedSpans ?? "0")) > 0n || partial.errorMessage)
    throw new Error(
      `OTLP partial success: ${String(partial.errorMessage ?? partial.rejectedSpans)}`,
    );
}
